plugins/ccgx/fu-ccgx-cyacd-file.c

```diff
@@ -0,0 +1,458 @@
+			version_row_num = row_num + version_row_index;
+		}
+
+		if (version_row_num == row_num) {
+			info->app_version.val= row_data [version_row_offset + 3] << 24 |
+							 row_data[version_row_offset + 2] << 16 |
+							 row_data[version_row_offset + 1] << 8 |
+							 row_data[version_row_offset + 0];
+			version_found = TRUE;
+		}
+
+		row_index++;
+
+		if (row_num == fw1_meta_row_num || row_num == fw2_meta_row_num) {
+
+			if (row_num == fw1_meta_row_num)
+				info->fw_mode = FW_MODE_FW1;
+			else
+				info->fw_mode = FW_MODE_FW2;
+
+			memcpy (&info->fw_metadata, &row_data [fw_meta_offset], sizeof(CCGxMetaData));
+
+			full_checksum = ((guint8)(1u) + (guint8)(~checksum));  /* the 2's complement of the binary sum. */
+			if (full_checksum != info->fw_metadata.fw_checksum) {
+				g_warning ("cyacd fw checksum error 0x%X != 0x%X", full_checksum, info->fw_metadata.fw_checksum);
+				break;
+			}
+
+			if (fw_size != info->fw_metadata.fw_size) {
+				g_warning ("cyacd fw size error %d != %d", (gint32)fw_size, (gint32)info->fw_metadata.fw_size);
+				break;
+			}
+
+			if (CCGX_METADATA_VALID_SIG != info->fw_metadata.metadata_valid) {
+				g_warning ("cyacd meta valid error 0x%02X != 0x%02X", (guint32)CCGX_METADATA_VALID_SIG, (guint32)info->fw_metadata.metadata_valid);
+				break;
+			}
+			fw_mode_found =	TRUE;
+
+			break;
+		}
+	}
+
+	if (version_found && fw_mode_found) {
+		/* set to row data start position */
+		fu_ccgx_cyacd_file_set_pos (handle, row_data_start_pos);
+		return TRUE;
+	}
+	return FALSE;
+}
+
+/**
+ * fu_ccgx_cyacd_file_read_row:
+ * @handle: Cyacd File Handle
+ * @data[in,out] cyacd data buffer
+ * @size Maximun size of cyacd data buffer
+ *
+ * Read row data in cyacd buffer
+ *
+ * Returns: %TRUE for success
+ */
+gboolean
+fu_ccgx_cyacd_file_read_row (CyacdFileHandle *handle, guint8 *data, guint32 size)
+{
+	guint32 read_size = 0;
+	g_autofree guint8 *ascii_buffer = g_malloc0 (CYACD_ROW_ASCII_BUFFER_SIZE);
+
+	g_return_val_if_fail (handle != NULL ,	FALSE);
+	g_return_val_if_fail (data != NULL , FALSE);
+	g_return_val_if_fail (ascii_buffer != NULL , FALSE);
+
+	read_size = cyacd_buffer_read_line (handle, ascii_buffer, CYACD_ROW_ASCII_BUFFER_SIZE);
+
+	if (read_size > 0) {
+		if (ascii_buffer[0] == ':' && ascii_buffer[1] == '0' && ascii_buffer[2] == '0') { /* : and 1byte array_id */
+			if (size > ((read_size - 5) / 2)) { /* 5 = 1 (:) + 2(array_id) + 2(checksum) */
+				size = (read_size - 5) / 2;
+			}
+
+			/* convert ASCII data in HEX data.
+			 * as per .cyacd format, first byte is :, second byte is
+			 * row ID and last is checksum. Ignoring these two enteries*/
+
+			convert_ascii_to_hex (&ascii_buffer[3], read_size - 3, data, size);
+
+			/* first two enteries : row_num and row_length are
+			 * stored in big endian format in file. wwap them to
+			 * little endian. */
+
+			*((guint16 *)&data[0]) = ((data[0] << 8) | (data[1]));
+			*((guint16 *)&data[2]) = ((data[2] << 8) | (data[3]));
+
```

Comment:
I think you want GUINT16_FROM_BE here

Reply:
I want to convert 16 bit big-endian to 16 bit little-endian explicitly because the device uses little-endian.
I am not sure that GUINT16_FROM_BE can convert guint16 value from big-endian to big-endian because it use host byte-order.
If yes, I can use like below. 
*((guint16 *)&data[0]) = GUINT16_FROM_BE ((data[0] << 8) | (data[1])